apps/emqx/src/emqx_persistent_session_ds.erl

```diff
@@ -845,124 +658,201 @@ do_ensure_all_iterators_closed(_DSSessionID) ->
+        case IsReplay of
+            true -> ItBegin0;
+            false -> ItEnd0
         end,
-        RangeIds
-    ).
+    case emqx_ds:next(?PERSISTENT_MESSAGE_DB, ItBegin, BatchSize) of
+        {ok, ItEnd, Messages} ->
+            {Inflight, LastSeqnoQos1, LastSeqnoQos2} = process_batch(
+                IsReplay, Session, ClientInfo, FirstSeqnoQos1, FirstSeqnoQos2, Messages, Inflight0
+            ),
+            Ifs = Ifs0#ifs{
+                it_begin = ItBegin,
+                it_end = ItEnd,
+                %% TODO: it should be possible to avoid calling
+                %% length here by diffing size of inflight before
+                %% and after inserting messages:
+                batch_size = length(Messages),
+                last_seqno_qos1 = LastSeqnoQos1,
+                last_seqno_qos2 = LastSeqnoQos2
+            },
+            {Ifs, Inflight};
+        {ok, end_of_stream} ->
+            %% No new messages; just update the end iterator:
+            {Ifs0#ifs{it_begin = ItBegin, it_end = end_of_stream, batch_size = 0}, Inflight0};
+        {error, _} when not IsReplay ->
+            ?SLOG(info, #{msg => "failed_to_fetch_batch", iterator => ItBegin}),
+            {Ifs0, Inflight0}
+    end.
 
-%% must be called inside a transaction
--spec session_drop_offsets(id()) -> ok.
-session_drop_offsets(DSSessionId) ->
-    OffsetIds = session_read_offsets(DSSessionId, write),
-    lists:foreach(
-        fun(OffsetId) ->
-            mnesia:delete(?SESSION_COMMITTED_OFFSET_TAB, OffsetId, write)
+%% key_of_iter(#{3 := #{3 := #{5 := K}}}) ->
+%%     K.
+
+process_batch(_IsReplay, _Session, _ClientInfo, LastSeqNoQos1, LastSeqNoQos2, [], Inflight) ->
+    {Inflight, LastSeqNoQos1, LastSeqNoQos2};
+process_batch(
+    IsReplay, Session, ClientInfo, FirstSeqNoQos1, FirstSeqNoQos2, [KV | Messages], Inflight0
+) ->
+    #{s := S, props := #{upgrade_qos := UpgradeQoS}} = Session,
+    {_DsMsgKey, Msg0 = #message{topic = Topic}} = KV,
+    Comm1 = emqx_persistent_session_ds_state:get_seqno(?committed(?QOS_1), S),
+    Comm2 = emqx_persistent_session_ds_state:get_seqno(?committed(?QOS_2), S),
+    Dup1 = emqx_persistent_session_ds_state:get_seqno(?dup(?QOS_1), S),
+    Dup2 = emqx_persistent_session_ds_state:get_seqno(?dup(?QOS_2), S),
+    Rec = emqx_persistent_session_ds_state:get_seqno(?rec, S),
+    Subs = emqx_persistent_session_ds_state:get_subscriptions(S),
+    Msgs = [
+        Msg
+     || SubMatch <- emqx_topic_gbt:matches(Topic, Subs, []),
+        Msg <- begin
+            #{props := SubOpts} = emqx_topic_gbt:get_record(SubMatch, Subs),
+            emqx_session:enrich_message(ClientInfo, Msg0, SubOpts, UpgradeQoS)
+        end
+    ],
+    {Inflight, LastSeqNoQos1, LastSeqNoQos2} = lists:foldl(
+        fun(Msg = #message{qos = Qos}, {Acc, SeqNoQos10, SeqNoQos20}) ->
+            case Qos of
+                ?QOS_0 ->
+                    SeqNoQos1 = SeqNoQos10,
+                    SeqNoQos2 = SeqNoQos20;
```

Comment:
Q: Admittedly, it's unclear to me why there's now 2 independent seqno tracks. Nice thing is `SeqNoRec - SeqNoComp` now gives something meaningful, but this property doesn't seem to be used (almost) anywhere. Or does it, do we plan to?

Reply:
With separate tracks we don't have to remember the QoS of the sent packets in `emqx_inflight` to check the correctness of PUBRELs. 
Otherwise we'd have to keep track of the QoS for each sent packet ID somewhere, that's extra memory and extra CPU.